skglm/penalties/separable.py

```diff
@@ -33,14 +33,17 @@ def prox_1d(self, value, stepsize, j):
     def subdiff_distance(self, w, grad, ws):
         """Compute distance of negative gradient to the subdifferential at w."""
         subdiff_dist = np.zeros_like(grad)
+        is_w_ws_provided = len(w) == len(ws)
```

Comment:
what about calling this function with `ws=np.arange(n_features)` as we currently do in other parts of the code?

Reply:
Yes indeed, but here what differs is that I pass in ``w_ws``  instead of ``w``

https://github.com/Badr-MOUFAD/skglm/blob/baea36a2fcc3ff5b6aaac308965a4179bb3cee74/skglm/solvers/pn_solver.py#L226

Reply:
in order to be 100% sure and clean, instead of creating `w_ws`, one could create a copy `w_copy` of the same size of `w`, and then  call `penalty.subdiff_distance(w_copy, cached_grads, ws)`. I think this is what we do in the rest of the code.

If this does not decrease performance, I would vote for this option.

Reply:
To avoid the creation of a potentially large array (`w_copy`), @Badr-MOUFAD would it work to pass the restricted w, the gradient, and  `ws=np.arange(len(restricted_w)` ?

I'm proposing this because I'm strongly in favor of leaving this function untouched and finding a workaround directly in the solver.

Reply:
Yes absolutely, it will work, (actually I initially implemented it this way)
However, this trick won't work if we are using a penalty that depends on the feature e.g. WeightedL1

Reply:
indeed, and such a bug will be nasty to catch.
So I'd go with the solution @QB3 proposed, instantiate a full array of shape  `n_features,`